lib/SILOptimizer/Analysis/SimplifyInstruction.cpp

```diff
@@ -298,11 +327,41 @@ SILValue InstSimplifier::visitAddressToPointerInst(AddressToPointerInst *ATPI) {
 
 SILValue InstSimplifier::visitPointerToAddressInst(PointerToAddressInst *PTAI) {
   // (pointer_to_address strict (address_to_pointer x)) -> x
+  //
   // If this address is not strict, then it cannot be replaced by an address
   // that may be strict.
-  if (auto *ATPI = dyn_cast<AddressToPointerInst>(PTAI->getOperand()))
-    if (ATPI->getOperand()->getType() == PTAI->getType() && PTAI->isStrict())
-      return ATPI->getOperand();
+  //
+  // When ownership is enabled, we can only perform this if the underlying
+  // address of our operand is either not an interior pointer or if it is an
+  // interior pointer if all uses of the pointer to address are with in the
+  // borrow scope that the interior pointer belongs to.
+  if (auto *ATPI = dyn_cast<AddressToPointerInst>(PTAI->getOperand())) {
+    if (ATPI->getOperand()->getType() == PTAI->getType() && PTAI->isStrict()) {
+      if (!PTAI->getFunction()->hasOwnership())
+        return ATPI->getOperand();
+
+      // If we do not have an interior pointer operand, we can just use this
+      // without issue. If we do have an interior pointer operand, we have to
+      // make sure that we are not adding a new interior pointer address to a
+      // borrow scope. This can happen since the address to pointer can be used
+      // to escape an address out of a borrow scope. If the pointer to address
+      // that rematerializes the address is outside the borrow scope and has
+      // uses outside of the borrow scope, we could introduce an ownership
+      // violation.
+      //
+      // TODO: We need to be able to exhaustively traverse all uses of the
+      // resulting address to determine if all are within the borrow scope. This
+      // requires using the access scope. Instead of doing that, we just bail
+      // here.
+      SILValue baseAddr =
+          getUnderlyingObjectStoppingAtMixedProjections(ATPI->getOperand());
```

Comment:
@atrick this is the issue I was telling you about with respect to address to pointer <-> pointer to address. I just tried to recognize if we had an interior pointer operand and bail in that case.